### PR TITLE
GEOMESA-723 Wire up TDI via WPS

### PR DESCRIPTION
**Pull request by kevinwheeler**
Signed-off-by: Kevin Wheeler <kevinwheeler90@gmail.com>

### CODE REVIEW

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/core.scala [hunk 1]
```diff
@@ -38,17 +38,18 @@ package object core {
...
+  val GEOMESA_ITERATORS_SFT_INDEX_VALUE          = "geomesa.iterators.sft.index-value-schema"
+  val GEOMESA_ITERATORS_ATTRIBUTE_NAME           = "geomesa.iterators.attribute.name"
+  val GEOMESA_ITERATORS_ATTRIBUTE_COVERED        = "geomesa.iterators.attribute.covered"
+  val GEOMESA_ITERATORS_ECQL_FILTER              = "geomesa.iterators.ecql-filter"
+  val GEOMESA_ITERATORS_TRANSFORM                = "geomesa.iterators.transform"
+  val GEOMESA_ITERATORS_TRANSFORM_SCHEMA         = "geomesa.iterators.transform.schema"
+  val GEOMESA_ITERATORS_IS_DENSITY_TYPE          = "geomesa.iterators.is-density-type"
+  val GEOMESA_ITERATORS_IS_TEMPORAL_DENSITY_TYPE = "geomesa.iterators.is-temporal-density-type"
+  val GEOMESA_ITERATORS_INDEX_SCHEMA             = "geomesa.iterators.index.schema"
+  val GEOMESA_ITERATORS_VERSION                  = "geomesa.iterators.version"
```
Comment (jdk2pq) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/core.scala]: was this change intentional?

Comment (kevinwheeler) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/core.scala]: Yea -- I added GEOMESA_ITERATORS_IS_TEMPORAL_DENSITY_TYPE and then had to add whitespace to the rest of the entries to line everything up on the equals sign.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala [hunk 2]
```diff
@@ -109,6 +115,18 @@ object TemporalDensityIterator extends Logging {
     new Interval(s, e)
   }
 
+  def createFeatureType(origFetType : SimpleFeatureType) = {
```
Comment (jdk2pq) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala]: Picky: to be consistent with how we've written most signatures, this should be {parameter: Type} without the space between the parameter and colon. I'd also called this `origFeatureType`

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala [hunk 3]
```diff
@@ -117,6 +135,19 @@ object TemporalDensityIterator extends Logging {
     resultTS
   }
 
+  private val df = DateTimeFormat.forPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
```
Comment (jdk2pq) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala]: remove ;

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala [hunk 4]
```diff
@@ -109,6 +115,18 @@ object TemporalDensityIterator extends Logging {
     new Interval(s, e)
   }
 
+  def createFeatureType(origFetType : SimpleFeatureType) = {
+    //Need a filler namespace, else geoserver throws nullptr exception for xml output
+    val (namespace, name) = buildTypeName(origFetType.getTypeName)
+    val outNamespace =
+      if (namespace == null){
```
Comment (jdk2pq) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala]: move up a line

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala [hunk 5]
```diff
@@ -117,6 +135,19 @@ object TemporalDensityIterator extends Logging {
     resultTS
   }
 
+  private val df = DateTimeFormat.forPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
+
+  def timeSeriesToJSON(ts : TimeSeries): String = {
+    val jsonMap = ts.toMap.map { case (k, v) => k.toString(df) -> v };
```
Comment (jdk2pq) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala]: remove ;

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala [hunk 6]
```diff
@@ -117,6 +135,19 @@ object TemporalDensityIterator extends Logging {
...
+    new JSONObject(jsonMap).toString()
+  }
+
+  def jsonToTimeSeries(ts : String): TimeSeries = {
+    val objMapper: ObjectMapper = new ObjectMapper();
```
Comment (jdk2pq) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIterator.scala]: remove ;

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/process/temporalDensity/TemporalDensityProcess.scala [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+               ): SimpleFeatureCollection = {
+
+    logger.info("Attempting Geomesa temporal density on type " + features.getClass.getName)
+
+    if(features.isInstanceOf[ReTypingFeatureCollection]){
+      logger.warn("WARNING: layer name in geoserver must match feature type name in geomesa")
+    }
+
+   val interval = new Interval(startDate.getTime, endDate.getTime)
```
Comment (jdk2pq) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/process/temporalDensity/TemporalDensityProcess.scala]: align

File: geomesa-core/src/test/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIteratorTest.scala [hunk 8]
```diff
@@ -74,14 +74,16 @@ class TemporalDensityIteratorTest extends Specification {
...
-        f
+
+    def decodeFeature(e: Array[_]): SimpleFeature = {
+      val f = builder.buildFeature(e(0).toString, e.asInstanceOf[Array[AnyRef]])
+      f.getUserData.put(Hints.USE_PROVIDED_FID, java.lang.Boolean.TRUE)
+      f.getUserData.put(Hints.PROVIDED_FID, e(0).toString)
+      f
     }
 
+    val features = encodedFeatures.map(decodeFeature)
```
Comment (jdk2pq) [inline geomesa-core/src/test/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIteratorTest.scala]: :+1:

File: geomesa-core/src/test/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIteratorTest.scala [hunk 9]
```diff
@@ -117,7 +128,7 @@ class TemporalDensityIteratorTest extends Specification {
       (iter must not).beNull
 
       iter.length should be lessThan 150
-      iter.length should be equalTo 1
+      iter.length mustEqual 1
```
Comment (jdk2pq) [inline geomesa-core/src/test/scala/org/locationtech/geomesa/core/iterators/TemporalDensityIteratorTest.scala]: This is fine (it's my preference) -- but "should be equalTo" and "mustEqual" are functionally equivalent.

File: geomesa-core/src/test/scala/org/locationtech/geomesa/core/process/temporaldensity/TemporalDensityProcessTest.scala [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+import org.apache.hadoop.io.Text
+import org.geotools.data.simple.SimpleFeatureStore
+import org.geotools.data.{DataStore, DataUtilities, Query}
+import org.geotools.factory.Hints
+import org.geotools.filter.text.ecql.ECQL
+import org.geotools.filter.visitor.ExtractBoundsFilterVisitor
+import org.junit.runner.RunWith
+import org.locationtech.geomesa.core.data._
+import org.locationtech.geomesa.core.index.{Constants, QueryHints}
+import org.locationtech.geomesa.core.iterators.TemporalDensityIterator.{decodeTimeSeries, TIME_SERIES, jsonToTimeSeries}
```
Comment (jdk2pq) [inline geomesa-core/src/test/scala/org/locationtech/geomesa/core/process/temporaldensity/TemporalDensityProcessTest.scala]: organize imports

File: geomesa-core/src/test/scala/org/locationtech/geomesa/core/process/temporaldensity/TemporalDensityProcessTest.scala [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+
+import scala.collection.JavaConversions._
+
+import org.joda.time.{DateTime, DateTimeZone}
+
+
+@RunWith(classOf[JUnitRunner])
+class TemporalDensityProcessTest extends Specification {
+
+  sequential
```
Comment (jdk2pq) [inline geomesa-core/src/test/scala/org/locationtech/geomesa/core/process/temporaldensity/TemporalDensityProcessTest.scala]: I don't think using "sequential" will achieve anything without using ">>" in your tests. @anthonyccri is this correct? I'm a bit fuzzy on the use of "in" vs ">>"

Comment (elahrvivaz) [inline geomesa-core/src/test/scala/org/locationtech/geomesa/core/process/temporaldensity/TemporalDensityProcessTest.scala]: I think it applies in both cases, although the exact way that specs2 determines order is still a mystery to me.

**Comment by elahrvivaz** [comment]
Looks good to me. Some minor formatting issues that @jdk2pq mostly identified.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/core.scala (hunk 1 above)
Comment (jdk2pq) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/core.scala]: Gotcha. :+1:

**Comment by jnh5y** [comment]
Merged as ab0c25c0ec55032331451b8b125c39d39fa783c2 on 1.5.
